overlord/configstate/configcore/kernel.go

```diff
@@ -0,0 +1,165 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/overlord/devicestate"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+const (
+	OptionKernelCmdlineAppend              = "system.kernel.cmdline-append"
+	OptionKernelDangerousCmdlineAppend     = "system.kernel.dangerous-cmdline-append"
+	CoreOptionKernelCmdlineAppend          = "core.system.kernel.cmdline-append"
+	CoreOptionKernelDangerousCmdlineAppend = "core.system.kernel.dangerous-cmdline-append"
+)
+
+func init() {
+	supportedConfigurations[CoreOptionKernelCmdlineAppend] = true
+	supportedConfigurations[CoreOptionKernelDangerousCmdlineAppend] = true
+}
+
+func changedKernelConfigs(c RunTransaction) []string {
+	changed := []string{}
+	for _, name := range c.Changes() {
+		// Note that we cannot just check the prefix as we have
+		// system.kernel.* options also defined in sysctl.go.
+		if name == CoreOptionKernelCmdlineAppend || name == CoreOptionKernelDangerousCmdlineAppend {
+			nameWithoutSnap := strings.SplitN(name, ".", 2)[1]
+			changed = append(changed, nameWithoutSnap)
+		}
+	}
+	return changed
+}
+
+func validateParamsAreAllowed(st *state.State, devCtx snapstate.DeviceContext, params []string) error {
+	// gd, err := devicestate.CurrentGadgetInfo(st, devCtx)
+	// if err != nil {
+	// 	return err
+	// }
+	// logger.Debugf("gadget data read from %s", gd.RootDir)
+	// TODO use gadgetdata to check against allowed values
+
+	return nil
+}
+
+func validateCmdlineExtra(c RunTransaction) error {
+	changed := changedKernelConfigs(c)
+	if len(changed) == 0 {
+		return nil
+	}
+
+	st := c.State()
+	st.Lock()
+	defer st.Unlock()
+	devCtx, err := devicestate.DeviceCtx(st, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	for _, opt := range changed {
+		cmdExtra, err := coreCfg(c, opt)
+		if err != nil {
+			return err
+		}
+
+		logger.Debugf("validating %s=%q", opt, cmdExtra)
+		params, err := osutil.KernelCommandLineSplit(cmdExtra)
+		if err != nil {
+			return err
+		}
+		if opt == OptionKernelCmdlineAppend {
+			// check against allowed values from gadget
+			if err := validateParamsAreAllowed(c.State(), devCtx, params); err != nil {
+				return fmt.Errorf("while validating params: %v", err)
+			}
+		} else { // OptionKernelDangerousCmdlineAppend
+			if devCtx.Model().Grade() != asserts.ModelDangerous {
+				return fmt.Errorf("cannot use %s for non-dangerous model", opt)
+			}
```

Comment:
want' the agreement to just ignore it instead?

Reply:
I see that this is mentioned in the spec but you added it, not sure at which point. I don't remember discussing it. Maybe @mvo5 remembers? The main questions is mostly about gadgets that have this in their defaults and are used across different grades?

Reply:
I don't remember if there's an easy way to distinguish whether we are applying default settings vs it's a configuration coming from the API?

Reply:
Yeah, ideally we would error only for the API one, but meanwhile we can just log the warning. Re: spec I think I added it because it seemed reasonable as explained, but I don't know if this concrete point was discussed, so I will just remove from the spec.

Reply:
not sure 100% but I think the default case the context has use-defaults set to true in it, see code in configstate.Configure and devicestate/firstboot.go

Reply:
Thanks for the pointers, it looks like with that information we could know if this comes from API or not, but I think that will need some changes to pass around the flag to the handler. I will add a TODO as I think that would be too much for this PR.